### DISCUSSION TITLE
Benching long-running/async (business) processes

### DISCUSSION THREAD

**Discussion by lcmgh** [bencherdev/bencher#470] [Q&A] [answered]
Ideally I'd like to know how long an overall (business) process that may involve other systems takes including the duration of each of the "sub steps".

Is that a use case bencher can cover with one of its adapters?

**Comment by epompeii** [comment, marked as answer]
Yes, this is a use case that Bencher can handle.

You will need to format your results in [Bencher Metric Format (BMF) JSON](https://bencher.dev/docs/reference/bencher-metric-format/).

Then you can either:

- Use [the JSON adapter](https://bencher.dev/docs/explanation/adapters/#-json) with the `bencher run` CLI subcommand (ie `--adapter json`)
- Use [the Bencher Report API](https://bencher.dev/docs/api/projects/reports/#post-v0projectsprojectreports) directly

I would probably model each business process/sub-process as a Benchmark. Likewise, I would create a Benchmark for any aggregates that you want to track.

For example, if you had Process A and Process B you would have three Benchmarks:

- Process A
- Process B
- Total Process

And if Process A had two sub-processes, Process A1 and Process A2, you would have two more Benchmarks (for a total of five):

- Process A
- Process A1
- Process A2
- Process B
- Total Process

You can use [the built-in Latency Measure](https://bencher.dev/docs/explanation/benchmarking/#measures) if you like, or you can use a custom Measure if there is a name/units that makes more sense for your domain.

**Reply by sagudev** [reply]
Are there any instructions/examples on how to use custom Measure in BMF?

**Reply by epompeii** [reply]
There is this section in the docs:

> The `benchmark_name` object can contain multiple Measure names, slugs, or UUIDs as keys. If the value specified is a name or slug and the Measure does not already exist, it will be created for you. However, if the value specified is a UUID then the Measure must already exist. In this example, `latency` is the slug for the built-in Latency Measure. Each Project by default has a Latency (ie `latency`) and Throughput (ie `throughput`) Measure, which are measured in `nanosecond (ns)` and `operations / second (ops/s)` respectively. The Measure object contains a Metric with up to three values: `value`, `lower_value`, and `upper_value`. The `lower_value` and `upper_value` values are optional.

https://bencher.dev/docs/how-to/track-custom-benchmarks/

Does that help? Is there anything you need me to explain further?

**Reply by sagudev** [reply]
So whatever is in key, that is a measure, but how can we then define units for our custom measures?

**Reply by epompeii** [reply]
If the Measure does not already exist, then yes what is in the key will be used for the name and slugified for the slug.

With that said, you can also just go into the Bencher Console or [use the `bencher` CLI](https://bencher.dev/docs/api/projects/measures/#post-v0projectsprojectmeasures) and create a new Measure for your Project. That is also how you set the units.

**Reply by sagudev** [reply]
Okay, thanks.
